plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixClient.java

```diff
@@ -213,6 +221,27 @@ public PhoenixClient(PhoenixConfig config, ConnectionFactory connectionFactory,
                 identifierMapping);
         this.configuration = new Configuration(false);
         getConnectionProperties(config).forEach((k, v) -> configuration.set((String) k, (String) v));
+        JdbcTypeHandle bigintTypeHandle = new JdbcTypeHandle(Types.BIGINT, Optional.of("bigint"), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty());
+        this.aggregateFunctionRewriter = new AggregateFunctionRewriter<>(
+                this::quoted,
+                ImmutableSet.<AggregateFunctionRule<JdbcExpression>>builder()
+                        .add(new ImplementCountAll(bigintTypeHandle))
+                        .add(new ImplementCount(bigintTypeHandle))
+                        .build());
+    }
+
+    @Override
+    public Optional<JdbcExpression> implementAggregation(ConnectorSession session, AggregateFunction aggregate, Map<String, ColumnHandle> assignments)
+    {
+        // TODO support complex ConnectorExpressions
+        return aggregateFunctionRewriter.rewrite(session, aggregate, assignments);
```

Comment:
(I could have sworn I had commented this somewhere else already... Ah found it, a closed PR...)

For Phoenix it would execute this query for every split. I did not see any automatic code that adds the final aggregation. So somewhere we have to add that.
BaseJdbc seems to make the assumption that there is exactly one split, which is true for all JDBC based connectors, except Phoenix.

Let's add the final aggregation part in BaseJdbc so that it all generally works for other JDBC based connectors (with special casing the one-split case)

Reply:
In other words... This cannot be handled purely in the connector, right? The plan needs to change higher up.

Reply:
@lhofhansl I am sorry sir, for closing the other PR, I messed it up :( 
Thanks, I see that we are getting counts per split but the final aggregation of the results from each split isn't happening and I was assuming that trino layer would have some logic to sum up the results. Is that a wrong assumption to have?
(Attached the results per split and explain plan). Let me see what change should be done in BaseJDBC
![image](https://user-images.githubusercontent.com/4153129/152922535-11b13d29-ec4f-4d0b-bb0e-35d93c224bb9.png)
 
![image](https://user-images.githubusercontent.com/4153129/152922588-556fbfbf-4fe4-494e-bfa1-a2c65f8a6329.png)

Reply:
When you said BaseJdbc, I am assuming you are referring to [BaseJdbcClient](https://github.com/trinodb/trino/blob/master/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java)

Reply:
Connectors are expected to only consume things fully which they can pushdown. In case of Phoenix aggregation shouldn't be consumed entirely.

Making change in BaseJdbcClient would mean that other connectors will end up with a partial aggregation on top in all cases.

Ideally we need some changes to `applyAggregation` to indicate whether a final aggregation needs to be performed or not.

Reply:
@hashhar Yeah, I think this needs a general solution, perhaps similar - but hopefully less hardcoded - to `supportsTopN`/`isTopNGuaranteed`, which also causes the overall plan to change.

In BaseJdbcClient we'd do the post work only in case that there are more than one splits. That way there will be no surprises in the future when other Jdbc connectors support multiple splits, which, I assume, would a scenario more common in the future.

Rewriting the plan more extensively in Phoenix might work. I'd fear about losing some of the performance improvements in that case.

So when the Hive/Iceberg connectors do aggregation pushdown, they also handle an final merge operations needed in the connector?

Reply:
There are no aggregation pushdowns in Hive/Iceberg since there is nothing to push down aggregations to.

However yes as a principle current architecture expects pushdown to be done by connectors - engine only participates in planning and optimisation. See also comment at https://github.com/trinodb/trino/issues/10974#issuecomment-1032460507

Reply:
@hashhar Sorry for some basic questions:
1. Trino knows that there are multiple splits (getSplits) in the query and also it's an aggregate plan (isAggregateEnabled). Wouldn't the trino layer be able to build the plan to aggregate the results across splits? I mean before aggregate pushdown implementation, trino query plan had 2 fragments (as attached) where the root fragment has a `Aggregate (FINAL)` node and it was doing the summation across splits right? But when we add the push down logic, I was assuming only the Fragment 1 would be modified to push down the aggregate and the root fragment 0 will remain same. Why is that not the case? How was a top level aggregate possible before pushdown and why not after pushdown? Please excuse if I am making wrong assumptions, new to this area.
![image](https://user-images.githubusercontent.com/4153129/153126005-423c9509-ebf5-4bed-aa36-d8ddd0593ef2.png)
2. 
> Ideally we need some changes to applyAggregation to indicate whether a final aggregation needs to be performed or not.
What kind of indication would be needed and who would make use of the indication. Can you help me understand how things work here?

Thanks @lhofhansl and @hashhar for patiently helping with PR review.

Reply:
For 1 - the definition of pushdown is that everything gets done by the connector. You indicate that to the engine by returning a non-empty value from the `ConnectorMetadata#applyXXX` methods (see their JavaDocs). The connector has told the engine that it'll consume the aggregation so the engine follows. This logic doesn't depend on the number of splits.

For 2 - some operations are modelled in the SPI as being able to get partially pushed down (e.g. topN). This is indicated by a boolean value being returned along with the pushdown result from the `applyTopN` method for example. That boolean indicates to engine that connector hasn't consumed the operation fully and that the engine still needs to add a final plan node.

Reply:
OK, I think I got the comments here now after looking spending time understanding the code and other Rules :). so the way topN works is by implementing isTopNGuaranted (this [slack thread](https://trinodb.slack.com/archives/CFLB9AMBN/p1608281212432700) also talked about a similar issue with topN, I did have some questions on that). In the similar fashion, would it be good to have some way of having a PartialAggregation pushdown and indicate that trino indeed needs to add an AggregationNode when there are splits are something of sort? 
If I understand what you are recommending in ConnectorMetadata#applyAggregation, is it by a way of adding some semantics in the preparedQuery so that the top level plan will add a summation? 
Interesting things! :)

Reply:
@ranganathg And I chatted offline, and he had a great idea:

We can look at the Phoenix query plan, and if there is a pushable aggregate, use a single split only. In that case that will likely be beneficial anyway, as Phoenix knows how to best schedule its resource to execute the aggregate.
I.e. if the connector decides it can push an aggregate it will also decide to do so with a single split.

Reply:
That makes sense and will solve this particular case. Thanks for coming up with this.

However we would eventually want this to be handled transparently since other connectors (Cassandra for example) can also benefit from the concept of partial aggregation pushdown.